File: ch10/whotofile.c
#include <wait.h>
#include <stdio.h>
#include <fcntl.h>
#include <stdlib.h>
#include <unistd.h>
#include<sys/stat.h>
#include <sys/types.h>

int main()
{
    int pid;
    int fd;

    printf("About to run who into a file\n");

    /* create a new process or quit */
    if ((pid = fork()) == -1)
    {
        perror("fork");
        exit(1);
    }
    /* child does the work */
    if (pid == 0)
    {
        close(1);
        fd = open("userlist",O_RDWR|O_CREAT|O_APPEND,S_IRUSR|S_IWUSR|S_IXUSR);
        execlp("who", "who", NULL);
        perror("execlp");
        exit(1);
    }
    /* parent waits then reports */
    if (pid != 0)
    {
        wait(NULL);
        printf("Done running who.  results in userlist\n");
    }
}

File: ch10/ch10_src/src/bin/whotofile.rs
use nix::libc::{c_char, creat};
use nix::unistd::{close, dup2, execvp, fork, ForkResult};
use nix::sys::wait::wait;
use std::error::Error;
use std::ffi::CString;

// WARNING!!! It can't work
fn main() -> Result<(), Box<dyn Error>> {
    println!("About to run who into a file");

    match unsafe { fork()? } {
        ForkResult::Parent {  .. } => {
            wait()?;
            println!("Done running who. Results in userlist");
        }
        ForkResult::Child => {
            let _ =  dup2(-1, 1) ; // Close standard output by duplicating to an invalid fd
            let fd = unsafe { creat("userlist".as_ptr() as *const u8 as *const c_char, 0o644) };
            let _ = dup2(fd, 1); // Redirect standard output to the file
            let _ = close(fd);
            let who_args = vec![CString::new("who").unwrap()];
            execvp(&CString::new("who").unwrap(), &who_args)?;
            println!("execvp failed");
        }
    }

    Ok(())
}